src/rovr_control/rovr_control/calibrate_field_coordinate_server.py

```diff
@@ -0,0 +1,98 @@
+
+import math
+
+import rclpy
+from rclpy.action import ActionClient, ActionServer
+from rclpy.action.client import ClientGoalHandle
+from rclpy.action.server import ServerGoalHandle
+from nav2_msgs.action import Spin
+from rovr_interfaces.action import CalibrateFieldCoordinates
+from rovr_interfaces.srv import ResetOdom
+from rclpy.node import Node
+from rclpy.task import Future
+
+class CalibrateFieldCoordinateServer(Node):
+    def __init__(self):
+        super().__init__("calibrate_field_action_server")
+        self._action_server = ActionServer(
+            self,
+            CalibrateFieldCoordinates,
+            "calibrate_field_coordinates",
+            self.execute_callback,
+            cancel_callback=self.cancel_callback,
+        )
+        self.cli_set_apriltag_odometry = self.create_client(ResetOdom, "resetOdom")
+        self.future_odom = Future()
+
+        self.cli_spin = ActionClient(self, Spin, "spin")
+        self.future_spin = Future()
+
+    def execute_callback(self, goal_handle: ServerGoalHandle):
+        """This method rotates until we can see apriltag(s) and then sets the map -> odom tf."""
+        self.get_logger().info("Executing goal...")
+        result = CalibrateFieldCoordinates.Result()
+        result.success = False
+        self.get_logger().info("Beginning search for apriltags")
+
+        if not self.cli_set_apriltag_odometry.wait_for_service(timeout_sec=1.0):
+            self.get_logger().error("Apriltag odom service not available")
+            return result
+
+        if not self.cli_spin.wait_for_server(timeout_sec=1.0):
+            self.get_logger().error("Nav2 service not available")
+            return result
+
+        for i in range(1, 3):
```

Comment:
In the original calibration method, the robot would keep turning in place until an Apriltag is seen. This whole for loop approach seems to change that. At first glance, this for loop seemed strange to me, but I guess it does have the benefit of failing at some point if there is legitimately no Apriltag, whereas the old method would have just kept spinning forever. So maybe this way is better? It does look a bit weird but that might just be me. I can test this new method on my desktop PC when I get a chance later in the week.

Reply:
This just does 2 180 degree turns then fails if none is found

Reply:
Doesn't it turn by 180 degrees in the first iteration and 360 degrees in the second iteration? Since your spin goal is i * math.pi

Reply:
I thought it turned to a heading,but if its distance then this is incorrect

Reply:
Yeah, I think it's distance

Reply:
Okay, I just fixed this